### ISSUE TITLE
jwt_authn: stats by provider or stats_prefix

### ISSUE THREAD

**Issue by sschepens** [envoyproxy/envoy#15405]
It would be great o have stats that distinguish allows/denies for each provider.
This could be done with a single filter or with multiple filters and a `stats_prefix` attribute which is kind of normal for filters already.

I can probably take this, what seems like the best approach?

**Comment by sschepens** [comment]
@qiwzhang there are a ton of `Verifier` impls and they are sort of nested which means we would have to propagate the filters `stats_prefix` and `scope` to all Verifiers so the can then pass them to their children.

Does this seem right? or is there another way?

**Comment by qiwzhang** [comment]
Yes, that is correct.  Instead of passing two parameters, you can pass a `StatFactory` object to hide them
```
class StatFactory {
public:
    Stat& create(absl::string_view provider) PURE;
};
```

**Comment by sschepens** [comment]
@qiwzhang i was thinking, if we add provider stats in `ProviderVerifierImpl` when using `allow_missing` or `allow_missing_or_failed` then we would actually generate `denied` stats which maybe are not actually denied because we allow failed or missing.

Whereas if we add a simple `stats_prefix` to the filter, we could get real allowed, denied stats.

Just a thought.

**Comment by qiwzhang** [comment]
Definitely, it is easy and simple to add provider prefixed `stats` in `filter` level.  But it doesn't have `provider` info when its code is called.  When it is called, it will be an aggregated result, e.g. for RequireAll case,  multiple providers may have verified. 

You raised a good point of `allowed` and `denied` for provider prefixed stats may not be accurate if `allow_missing` or `allow_missing_or_failed` is used.   How about named the stats as `PROVIDER_verification_passed` and PROVIDER_verification_failed` counters.

**Comment by sschepens** [comment]
> Definitely, it is easy and simple to add provider prefixed stats in filter level. But it doesn't have provider info when its code is called. When it is called, it will be an aggregated result, e.g. for RequireAll case, multiple providers may have verified.

For our use case it would probably be OK and simpler to just add a prefixed stat at a filter level and just two filters one for each provider. But if anyone think it's worth pursuing the second option I could ahead and do that too.

> You raised a good point of allowed and denied for provider prefixed stats may not be accurate if allow_missing or allow_missing_or_failed is used. How about named the stats as PROVIDER_verification_passed and PROVIDER_verification_failed` counters.

Sure, seems more logical, do you think we should pursue this option or the filter prefixed stats?

**Comment by qiwzhang** [comment]
Let us go with the latter.  Thanks

**Comment by sschepens** [comment]
@qiwzhang one more thing, regarding what i raised last week, couldn't there be issues if a provider is specified more than once in the requirements? That would mean we have several `ProviderVerifierImpl` each one with it's own struct of stats but having the same stat names.

Not sure how envoy stats handle this cases.

**Comment by qiwzhang** [comment]
That is a good point.  Then, provider prefixed stat should not be constructed at `ProviderVerifierImpl`, it should be at filter_config.  [here](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/extensions/filters/http/jwt_authn/filter_config.h#L81)

then how do we pass it to authenticator to update it?

maybe pass it as callback interface like [setPayload](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/extensions/filters/http/jwt_authn/verifier.h#L35)

**Comment by sschepens** [comment]
@qiwzhang should we then store the stats in a map or is there another way to the identify the fields in the struct?

**Comment by qiwzhang** [comment]
Yes,  it will be faster if we use map.

We can define a new Stats struct just like [this](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/extensions/filters/http/jwt_authn/filter_config.h#L53) for each provider prefixed stat and loop through all providers to create the map
